Feature/various board sizes

serpient opened the pull request:
Closes #14  #4 #15 

![gif](https://media.giphy.com/media/Sx4HYz8W6cQBoDuTHY/giphy.gif)

# New Entry Point `TicTacToe.game_start()`

# Things Done
- added function to generate 3x3 or 4x4 (or any size really) board data and printable board string
- Large refactor to pass around a Board_Spec (struct holding board_data, num_of_columns, and num_of_rows) to all the functions instead of originally passing Board_Data.
- Integration of custom board size feature with the other Tic Tac Toe functions like checking for wins, board to row conversions, etc.
- Get user input and updating game settings before game play starts. 
- update input validation min/max range so its not a hard coded 1-9. use 1 and MAX number saved in BoardSpec

# Things done after Feedback
- Clean up board printer with better naming, and extracting implementing details in hopes of presenting a clearer order of operations
- Extract board struct into its own module `BoardState` with methods that retrieves and calculates values from the board data. This was done in realization that I have very tightly coupled functions -- there are numerous functions everywhere that were restructuring / pattern matching against BoardState, knowing that it was a `struct` data type. With a separate module, if I were to change the data structure or naming of my board state values, the functions used by everything else shouldn't have to change.
- Use atoms such as `:empty`, `:player`, and `:opponent` instead of hard coded `X, O, " "`. Decouples the string printing from the actual values
- Refactor functions to assume a single determination of board size. Before, I had assumed a `b * c` type board. Confirmed with mentors that its a `b * b` board, which cleaned up a few things. 

## Cool things to me
Board generator can do any board size, for example 
### a 6x6:
<img width="504" alt="Screen Shot 2019-04-09 at 1 34 25 PM" src="https://user-images.githubusercontent.com/29721784/55834124-0549db80-5ace-11e9-8b11-c92dd90e757c.png">

### or even a 11x11
<img width="489" alt="Screen Shot 2019-04-09 at 1 48 40 PM" src="https://user-images.githubusercontent.com/29721784/55834206-33c7b680-5ace-11e9-8ee9-6eee312e830e.png">

serpient committed:
generate_board to output custom board data

serpient committed:
write test for generate_board_for_print

serpient committed:
implement basic generate_board_for_print

- Takes custom row and column arguments
- takes board_data argument
- converts the above arguments into a printable board

serpient committed:
inject board data into board print

serpient committed:
refactor board printer; isolate cell printers

serpient committed:
adjust string print for double digit cell num

* fixes bug where it was printing misaligned for double digit numbers
* mix format

serpient committed:
print 3x3 board with generate_board_for_print

serpient committed:
add test unit for 3x4 board generation

serpient committed:
refactor board print using zip

- still not happy with how complicated it looks. mostly, theres a lot of conversions from tuple to list to tuple, back to list!

serpient committed:
generate cell by list instead of tuple->list

serpient committed:
expose top of pipe , add anonymous function to append margin

serpient committed:
reverse conditional for better flow; save string result to board; concat margin

serpient commented on lib/board.ex:
```diff
@@ -36,4 +36,54 @@ defmodule Board do
+    end
+  end
+
+  def compile_row_of_cells(num_of_columns, num_of_rows, row_idx, board_data) do
```
messy board print creation?

cmvandrevala replied:
Yeah, this is a bit tough to follow. Maybe rename some functions to things like "draw_symbol" or "horizontal_line" or "vertical_line"

serpient committed:
use a shorter map access method

serpient committed:
add default values for board generation

serpient committed:
remove unused get_board fn; replace with new one

serpient committed:
delete unused code; move 3x3 test print to board_test

serpient committed:
add test unit to retrieve a new board struct

serpient committed:
pass update_board_spec test unit

serpient committed:
refactor fn to parse board data from board spec

serpient committed:
name arg from board_data to board_spec for clarity

serpient committed:
refactor convert-to-row fns to take in board_spec

serpient committed:
add test unit to convert 4x4 correctly to horizontal

serpient committed:
add test unit for 4x4 vertical conversion

serpient committed:
convert_vertical_to_row passes for 4x4

serpient committed:
mix format

serpient committed:
add test watch setting

serpient committed:
refactor convert_vertical_to_row; fix bug

serpient committed:
add 4 board win test units for horiz/diag check

serpient committed:
add test unit for 4x4 diagonal conversion

serpient committed:
passing 4x4 diagonal conversion fn

serpient committed:
create take_every fn to handle conversion

serpient committed:
remove reference board comments

serpient committed:
add 3x3 generate_board_data test unit (default case)

serpient committed:
add passing diagonal wins check

serpient committed:
rename params for clarity

serpient committed:
update invalid_input_range message to not depend on 1-9

serpient committed:
add failing test unit to get game settings input

serpient committed:
extract compilation of cell parts for clarity

serpient committed:
format board file

serpient committed:
outputs a message to get board size input

serpient committed:
add game setting message and starting banner print

serpient committed:
update processInput fns, remove unused args

serpient committed:
update test unit with banner string output

serpient committed:
fix board_data update bug

- it was still updating with board_data as argument. needed to change it to board_spec (struct)

serpient committed:
fix validate input bug

- again, it was assuming board_data as argument, but needed to update it to board_spec (struct) type.
- also set the max input dynamically, instead of a hard code 9

serpient committed:
update test units for board and process input

- use board_spec instead of board_Data

serpient committed:
add clear_io to win and tie messages; update tests

serpient committed:
add game_start fn and link to play

serpient committed:
comment out game_start fn -- untestable at its current state

serpient committed:
formatting of banner concat

lpenzey commented on lib/check_for_wins.ex:
```diff
@@ -4,36 +4,39 @@ defmodule CheckForWins do
-  def analyze(board) do
-    horizontal = Board.convert_horizontal_to_row(board) |> is_any_row_within_board_a_win?
+  def analyze(board_spec) do
+    board = board_spec.board_data
```
One thing my mentors mentioned to me was to use getters/setters for accessing data within structs. As you have it here, the check_for_wins modules knows that board_spec responds to .board_data. If you encapsulated that access in a getter method within the Board module you could protect other modules from depending on the exact syntax of your struct fields

cmvandrevala replied:
@lpenzey You are definitely correct in an OO paradigm since data and behaviour are packaged together in a class. In a functional paradigm though, it is not uncommon to have structs store data and functions act on structs (no getters or setters)

lpenzey commented on lib/board.ex:
```diff
@@ -1,13 +1,34 @@
     end
   end
 
+  def update_board_data(position_to_update, board_spec, player_symbol) do
```
More of stylistic preference, but since this is namespaced within the `Board` module you could name it `update_data` since the call `Board.update_data` tells the full story

cmvandrevala replied:
Definitely agree ^

serpient committed:
refactor -- rename; move around pipe fns in board

serpient committed:
refactor -- rename; move around pipe fns in board

christophermanahan commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+    end)
+  end
+
+  def generate_board_for_print(board_spec) do
```
Should a board know how to print itself? I'm thinking single responsibility principle here. Maybe another class can be responsible for presentation logic?

cmvandrevala replied:
^ Definitely agree. Some kind of board printing module would be good

cmvandrevala commented on test/tic_tac_toe_test.exs:
```diff
@@ -41,13 +45,21 @@ defmodule Tic_Tac_Toe_Test do
+             capture_io(fn -> GameIO.print_win(winning_board_spec, "X") end)
   end
+
+  # test "[handle_game_start] Gets game settings, updates default board data, and starts game" do
```
You can remove this comment

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
 defmodule Board do
-  def handle_board_update(result, board_data, player_symbol) do
+  defstruct num_of_rows: 3,
+            num_of_columns: 3,
```
I would be careful putting `number_of_rows`, `number_of_columns`, and `max_spaces` in the struct because `max_spaces` is calculated from the other two. There is an opportunity for a developer to accidentally make a math error and introduce bugs. I would maybe just have a function to calculate it where needed?

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+  defstruct num_of_rows: 3,
+            num_of_columns: 3,
+            board_data: %{
+              1 => " ",
```
Instead of storing an empty space as an empty string, why not use something more aligned with the domain language? For example, why not use the atom `:empty`?

serpient replied:
Makes sense -- updated

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
     board
     |> Map.values()
-    |> Enum.chunk_every(3)
+    |> Enum.chunk_every(num_of_columns)
```
This is cool. I have never seen the use of chunk_every.

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+
+    1..num_of_rows
+    |> Enum.map(fn row_idx ->
+      take_board_value_by_every_num(row_idx..max_spaces, columns, columns, board)
```
It is a bit weird that you are passing columns in as an argument twice here.

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+  def convert_diagonal_to_row(board_spec) do
+    %Board{board_data: board, num_of_columns: columns, max_spaces: max_spaces} = board_spec
+
+    [1, columns]
```
The number 1 in this code is a "magic number". I really don't know where it is coming from. Maybe put it in a variable or constant?

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+    |> Enum.map(fn row -> board[row] end)
+  end
+
+  def generate_board_data(num_of_rows \\ 3, num_of_columns \\ 3) do
```
Be careful here. You specify the default number of rows and columns as three, but it really depends on the values set in the `Board` struct

cmvandrevala commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+    }
+
+    cond do
+      at_end_of_cell_line? == false -> line[cell_level] <> "|"
```
`at_end_of_cell_line? == false` can be re-written as `!at_end_of_cell_line?`

cmvandrevala commented on lib/game_IO.ex:
```diff
@@ -50,9 +41,40 @@ defmodule GameIO do
   end
+
+  def game_start_banner(message) do
+    padding_top = "\n"
```
Instead of declaring everything in variables and then concatenating the strings, why not just concatenate the raw strings?

cmvandrevala commented on lib/process_input.ex:
```diff
@@ -1,34 +1,35 @@
   end
 
-  defp process_input(string, _board, _current_player) when is_tuple(string) do
+  def process_input(string) when is_tuple(string) do
```
I might think of a better name than "process_input". Not sure what the function does, to be honest.

cmvandrevala commented on lib/tic_tac_toe.ex:
```diff
@@ -27,25 +27,28 @@ defmodule TicTacToe do
+
+    Board.generate_board_data(board_size, board_size)
+    |> Board.update_board_spec(board_size, board_size)
+    |> play(nil, "X")
```
What does `play(nil, "X")` do? That seems a bit strange to me.

serpient replied:
passing in nil was a lazy way of using the default message in the function. Replaced with `:initial_player_prompt`!

cmvandrevala commented on test/board_test.exs:
```diff
@@ -1,5 +1,6 @@
 defmodule Board_Test do
   use ExUnit.Case
+  import ExUnit.CaptureIO
```
This might be ok for now, but we will talk about how to handle IO, databases, and other hard-to-test parts of code in future workshops.

serpient replied:
I'm looking forward to it !

cmvandrevala reviewed (commented):
Your TTT game looks pretty good. The one thing that I would recommend is to try to use as much domain-specific vocabulary as you can - especially in your tests. For example, you have a number of tests that have lines similar to this: `error_board_spec = Board.update_board_spec(board_with_error_input)`. I am not quite sure what they are doing in the tests. But something like `new_board = Board.move(:player_one, "X", 5)` is easier to interpret

cmvandrevala commented:
Also, one other thought - admittedly your board generator is super cool, and I think your console printed boards are some of the prettiest I have seen. But be careful adding in features that you don't necessarily need. I am guessing that your mentor did not require 11x11 boards, but if you introduce that feature to a client, you have to support it for perpetuity. That might not seem like a big deal until someone asks for an unbeatable computer player to play TTT - you have to create a computer player that can play optimally on any board size, not just a 3x3 board.

serpient commented:
Thanks for the feedback @cmvandrevala  @christophermanahan  @lpenzey !! I couldn't see a way forward, so all the feedback will be VERY helpful as I refactor this tomorrow.

I really appreciate it! 

![gif](https://media.giphy.com/media/3oz8xIsloV7zOmt81G/giphy.gif)

serpient commented:
> Also, one other thought - admittedly your board generator is super cool, and I think your console printed boards are some of the prettiest I have seen. But be careful adding in features that you don't necessarily need. I am guessing that your mentor did not require 11x11 boards, but if you introduce that feature to a client, you have to support it for perpetuity. That might not seem like a big deal until someone asks for an unbeatable computer player to play TTT - you have to create a computer player that can play optimally on any board size, not just a 3x3 board.

I understand!! It was a fun little thing, but I didn't think of the repercussions of having to use minimax on a board larger than 4x4. I had already heard a 4x4 board can slow down the function alot, but didn't connect it to this. My game start message asks for either a 3x3 or 4x4 board for now, so I can add a input validation to make sure it doesn't generate a bigger game.

ssunday commented on lib/board.ex:
```diff
@@ -1,39 +1,148 @@
+
+    %Board{
+      existing_board
+      | board_data: new_board_data,
```
Maybe extract out the updated data out so you can existing_board | updated_board_data?

ssunday commented on lib/check_for_wins.ex:
```diff
@@ -4,36 +4,39 @@ defmodule CheckForWins do
+      Enum.map(row_data, fn row -> is_row_a_win?(row, "X") || is_row_a_win?(row, "O") end)
 
     cond do
       Enum.member?(wins_list, true) -> true
```
Couldn't any be used for this?

serpient replied:
Yes, you are right! I've updated 💃

serpient committed:
extract board print functions to separate module

serpient committed:
remove comment

serpient committed:
change fn name to new_struct for clarity

serpient committed:
mix format

serpient committed:
extract board print tests

serpient committed:
remove default in board data generator; format

serpient committed:
reformat board printer; rename and extract fn

ssunday commented on lib/board.ex:
```diff
@@ -1,39 +1,105 @@
-        row_idx == 1 -> [board[row_idx], board[row_idx + 4], board[row_idx + 8]]
-        row_idx == 3 -> [board[row_idx], board[row_idx + 2], board[row_idx + 4]]
+  def convert_diagonal_to_row(board_spec) do
+    %Board{board_data: board, num_of_columns: columns, max_spaces: max_spaces} = board_spec
```
`board_spec`....I think I get what it means, but it feels odd to me

serpient replied:
what do you think of any of the following:
`board`
`board_attributes`
`board_state`
?

ssunday replied:
`board_state` IMO

serpient replied:
Will do! Updating to `board_state`

serpient committed:
refactor to assume symmetric board size

- I created the code on assuming a n * m board size.
- Need to refactor to assume a n * n board size.

serpient committed:
rename play to move()

serpient committed:
add descriptive variables

serpient committed:
reduce variables in banner message printing

ssunday commented on lib/game_IO.ex:
```diff
@@ -50,9 +41,33 @@ defmodule GameIO do
+    ~s(               T I C . T A C . T O E                         \n\n) <>
+    ~s(  #{GameIO.get_message(message)}             \n) <>
+    String.duplicate(empty_row, 2) <>
+    ~s(#######################################################     \n) <>
```
Could you pull out the dividers into variables or some constant so that you don't repeat them visually?

serpient committed:
change fn to use Enum.ANY instead

serpient committed:
use variable for border row strings

serpient committed:
extract board state into its own module

- the other fn knew too much about the type of board state (map) as it was accessing values specifically for maps everywhere.
- Sp, extract all the state into its own module and have individual getter methods

serpient committed:
decouple structure logic more

- extract map.values to its own boardState fn to decouple knowledge of map data structure
- rename move to play
- rename handle fn to handle_play for clarity

serpient committed:
rename print_board to string_board

serpient committed:
rename print_board to printable_board

serpient committed:
merge conflict

serpient committed:
fix merge conflict

ssunday commented on lib/board_state.ex:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def new_state(new_board_data, new_board_size \\ 3) do
+    state = %BoardState{}
```
why do you need to initialize state if it's empty?

serpient replied:
ur right! removing

serpient committed:
set " " to :empty,  X / O to :opponent & :player

- decouple from hardcoded string

ssunday commented on lib/board.ex:
```diff
@@ -11,11 +11,11 @@ defmodule Board do
     board_state
     |> BoardState.all_keys()
-    |> Enum.any?(fn value -> value == " " end)
+    |> Enum.any?(fn value -> value == :empty end)
```
woo

serpient committed:
update atom use instead of hard strings; mix format

serpient committed:
rename process_input

serpient committed:
set implementation fns private

serpient committed:
remove unnecessary board struct declaration

aaizuss commented on lib/board.ex:
```diff
@@ -1,39 +1,62 @@
-    |> Enum.any?(fn value -> value == " " end)
+  def has_empty_spaces?(board_state) do
+    board_state
+    |> BoardState.all_keys()
```
what does all_keys mean? I assume this is getting all of the positions, or spaces (you use position earlier, so maybe call it position to be consistent)

aaizuss commented on lib/board.ex:
```diff
@@ -1,39 +1,62 @@
+    |> Enum.any?(fn value -> value == :empty end)
   end
 
   def is_a_empty_space(board_position, board) do
```
sometimes you use the term `space`, and sometimes you use `position`. Let's pick one and be consistent. I like that you're being descriptive with the function name here, but I think you can drop the `a` and call it `is_empty` or `is_space_empty`

aaizuss commented on lib/check_for_wins.ex:
```diff
@@ -4,32 +4,32 @@ defmodule CheckForWins do
   end
 
   def is_any_row_within_board_a_win?(all_rows) do
```
Does the name `has_winning_row` also communicate what this function does?

aaizuss commented on lib/check_for_wins.ex:
```diff
@@ -4,32 +4,32 @@ defmodule CheckForWins do
-  def analyze(board) do
-    horizontal = Board.convert_horizontal_to_row(board) |> is_any_row_within_board_a_win?
+  def analyze(board_state) do
+    horizontal = Board.convert_horizontal_to_row(board_state) |> is_any_row_within_board_a_win?
```
`convert_horizontal_to_row` alters the data structure for the board so you can determine if there's a winner. This line might be clearer if you rename that function to `horizontal_rows`. Then you'd have

`horizontal = Board.horizontal_rows(board_state) |> has_winning_row?`

I think we can clarify the variable name horizontal as well - it's referring to whether there are horizontal wins, right? So `horizontal_wins` should be fine because trues in the list would refer to wins

aaizuss commented on lib/check_for_wins.ex:
```diff
@@ -4,32 +4,32 @@ defmodule CheckForWins do
-      Board.has_empty_spaces?(board) == false -> {:error, :game_is_a_tie, board}
-      true -> {:ok, :no_win, board}
+      Enum.any?(board_has_win?) -> {:ok, :wins_game, board_state}
+      Board.has_empty_spaces?(board_state) == false -> {:error, :game_is_a_tie, board_state}
```
I try not to explicitly compare with boolean values. You could nagate `Board.has_empty_spaces` here, but that might also be unclear, so I'd recommend adding an `is_full?` function to your Board module. And if this is the only place where you use `has_empty_spaces`, then I'd recommend removing this function and replacing it with the is_full function because it's clearer for this use case.

aaizuss commented on lib/check_for_wins.ex:
```diff
@@ -4,32 +4,32 @@ defmodule CheckForWins do
   end
 
-  def check_for_win(result, board, _current_player) do
+  def check_for_win(result, board_state, _current_player) do
```
I think you can remove this.

aaizuss commented on lib/printable_board.ex:
```diff
@@ -0,0 +1,76 @@
+    |> List.insert_at(-1, "\n\n\n")
+  end
+
+  def generate_board_for_print(board_state) do
```
Since this is the function used outside of the module, it would be more readable for a developer to put this function at the top of this module, and then order the rest of the functions below it based on whether they're high/low level helpers

aaizuss commented on lib/printable_board.ex:
```diff
@@ -0,0 +1,76 @@
+  defp create_row_of_cells(board_size, row_idx, at_bottom_of_last_row?, board_data) do
+    1..board_size
+    |> Enum.map(fn column_idx ->
+      curr_column_number = row_idx * board_size + column_idx
```
I was confused at first about the difference between a column_index and a column_number. `curr_column_number` refers to the number the user sees, so I think `current_cell_display_number` would be clearer

aaizuss commented on lib/process_input.ex:
```diff
@@ -1,34 +1,37 @@
 defmodule ProcessInput do
-  defp transform_to_integer(string) do
+  def transform_to_integer(string) do
```
Ultimately, this is converting user input, right? Can you make this name more specific?

aaizuss replied:
Also, if you're always calling `handle_integer_conversion_result` after `transform_to_integer`, you can probably combine them into one function. I don't think `transform_to_integer` is a useful abstraction

serpient committed:
rename fns for clarity; refactor per comments
